### ISSUE TITLE
Recap of All Issues

### ISSUE THREAD

**Issue by Pelonza** [Pelonza/multivariable-calculus-IBL#24]
@davidfarmer

Things still going wrong:
Hints are not rendered within the knowl
Tikz pictures are not being created/rendered
--Issue with cross-referencing problems now -- (maybe exploration not the best environment..hrm.)
Render the macro \wolfA as a hyper-link to WolframAlpha

After conversion to PTX: (according to your response)
Creation/conversion of textbook references to a "Additional Resources" knowl
Any NOT textbook references, converted to an "Insight" knowl

-------------
Generally... this is fairly close, and I'm going to have to do a careful read-through anyway to actually fix/add all the rest of the content I want... so these _could_ be ok. If I understood how to fix the latex-image rendering etc....

**Comment by Pelonza** [comment]
Re: Images...

I didn't realize I needed to separately run the MBX script from mathbook to render the images or the sage plots... so I'm going to try that now...

**Comment by davidfarmer** [comment]
If you want hints in a knowl, that is done with a flag when you compile.

You already figured out that there is a separate step to make the images.

Not sure about references to explorations.  If something is numbered,
then it should be possible to reference it.  But maybe there is a bug
in the PTX code?

\wolfA will be handled later.

So maybe you are ready to have the PTX files as "official"?

**Comment by Pelonza** [comment]
So, there's two things here...

1) I think there's an error somewhere in the PTX code... because hints
CREATES a knowl... but if you look at them, the TEXT of the hint is NOT
inside the knowl, but rather after where the knowl appears. I don't _think_
this is actually a problem with the conversion (the .mbx file looks
"correct" as I understand it)

2) I think you are correct that, overall I think these could be considered
"official" ptx files. It seems that the other conversions of macros can
probably be handled with another parsing script, although if your script
correctly converts the open/close '{}' for a macro, perhaps the script
should replace the "insight" references with some sort of markdown style
thing? (even if it's a 'nonsense' markdown) That would make it easier to
find/replace later, rather than making sure I correctly get all the {}'s....

But I'm willing to take advice on that since I suspect I'll need to
carefully read through each chapter/section anyway to add
sage/webworks...so can hopefully fix everythign as I do that.



==================================================
To schedule a meeting or appointment try:
https://karlrbschmitt.youcanbook.me/

Dr. Karl Schmitt
Assistant Professor
Department of Mathematics and Statistics
Department of Computing and Information Sciences
Director of Data Science Program
Director of Analytics and Modeling Graduate Program
Valparaiso University, Indiana
==================================================

**Comment by davidfarmer** [comment]
Do you ahve an updated HTML illustrating the issue with hint
knowls?

If examining that indicated the source is good but the HTML is bad,
then we can submit an PTX issue and proceed using your PTX files.

**Comment by Pelonza** [comment]
The same website as before has it up...

http://faculty.valpo.edu/calculus3ibl/index.html

Problem 3.6.4 has an example of a hint incorrectly rendered. But, as I read
the mbx (you should confirm) it seems to have been written correctly in ptx.

And justto se it wasn't JUST there, you can see it again in 5.1.9

**Comment by davidfarmer** [comment]
I see a bunch of unprocessed macros:

\bmw
\valposhort
\instructor

I suggest trying to fix those in the PTX.

**Comment by Pelonza** [comment]
Ok. If you are able to easily replace macro instances with a PTX
environment, the "\bmw", "\valpo" and "\valposhort" are easy, and should go
into an "insight" environment. Otherwise, I can try to catch them all by
hand and replace them.

The \instructor macro is a lot harder, and isn't going to be easily fixed
in PTX either... that's related to providing a "switch" or "instructor
only" comments... which I'm not aware of being available.

**Comment by davidfarmer** [comment]
A reasonable (but possibly temporary) way to handle things like
instructor switches, is to invent a new tag, such as < instructor >.

Then you are faced with the problem of how to hide the contents
of that tag by default (this should be a relatively simple, once
one knows a little xslt), or to do something useful with it when
needed (probably just slightly more difficult).

I can put "\bmw", "\valpo" and "\valposhort" into an "insight"
environment, with the argument of "\section" as the title of
the "insight".  I'll make a pull request, but not sure I can get to it
tomorrow.

**Comment by Pelonza** [comment]
Thanks for the continued work on this David! And for the update on timing.

Can you also include the macro:

"\marginparbmw" as an "insight"

So in total, 4 macros into an "insight" environment, titled with the
section (or current exploration?) :
\bmw
\marginparbmw
\valpo
\valposhort

The macros:
\instructor
\notes

Can just be commented out in the PTX. It's not that frequently used, and
can probably just be collected as an appendix for instructors (I don't
think there's any "hidden" information, just unrelevant to student info)

Similarly the macros:
\thomasee
\thomaseeshort
\larsonfive
\larsonfiveshort
\stewarts
\stewartsshort

Can also just be commented out. At this point they are all out-dated
references (new editions of textbooks), and with the web, better things can
be linked/helped I think (if I want to do that more).

----------

I think with all those macro's taken care of, the PTX files should be fine
for being finalized. The \wolfA I can simply do a find/replace throughout
the files easily enough. Everything else is just editing for me I think...
(except getting the "hints" to render properly, but I don't _think_ that's
an issue with the ptx markdown, but rather the ptx style-sheet/processing.

**Comment by davidfarmer** [comment]
I am working on the conversion to "insight" right now,
but for some reason my prettyprinter is adding spaces in
other locations.  Should have something today.

**Comment by Pelonza** [comment]
@davidfarmer 

I just did your latest pull request. In the comment you said you only converted 3 of the four macros to insights?

Did you also convert "\marginparbmw" ?

The "\instructor" and "\note" can just be commented out, but included in the PTX documents.

**Comment by Pelonza** [comment]
@davidfarmer 

New Issue: All the problem content for the explorations needs to be wrapped in a "statement" syntax.

New Issue: The section "review_matrices" is still using "project" instead of "exploration" (is anything else?)

Given the scope of this, I don't want to try and fix it by hand...

**Comment by Pelonza** [comment]
@davidfarmer 

New Issues: All the problem content needs to be wrapped in a "statement" markdown, per the thread on the support-email list. Apparently that's why the hints don't render properly.

New Issue: in the "review_matrices.mbx" it's still using "project" not "exploration" 

New Issue: mbx files won't compile to html:
../ptx/chapter1.mbx:29: parser error : Opening and ending tag mismatch: insight line 26 and li
            </li>
                 ^
../ptx/chapter1.mbx:35: parser error : Opening and ending tag mismatch: li line 31 and insight
            </insight>
                      ^
./../ptx/index.mbx:117: element include: XInclude error : could not load ../ptx/chapter1.mbx, and no fallback was found
../ptx/section11-2.mbx:243: parser error : Opening and ending tag mismatch: insight line 240 and li
            </li>
                 ^
../ptx/section11-2.mbx:255: parser error : Opening and ending tag mismatch: li line 251 and insight
              </insight>
                        ^
../ptx/chapter11.mbx:308: element include: XInclude error : could not load ../ptx/section11-2.mbx, and no fallback was found


-------
I'm assuming these are symptoms of other issues....but perhaps not.

**Comment by Pelonza** [comment]
RE: Won't compile to html...

I fixed the two locations where the insight seems to have been inserted incorrectly. And now it compiles...

**Comment by Pelonza** [comment]
But now some of the other macro's seem to have broken:

http://faculty.valpo.edu/calculus3ibl/section-14.html
See definition... "\proj" not rendering.
Similar, http://faculty.valpo.edu/calculus3ibl/section-7.html
the "\amp" is rendering wrong.... in a matrix.

**Comment by davidfarmer** [comment]
See if you can fix the parser errors by hand.  Those are probably due to
the nesting problem I mentioned.  (You will find that some of the insights
are commented out, which may not be what you want.)

I don't see the word "project" in the file review_matrices.mbx , so I am
not sure why that is appearing in the html.

**Comment by davidfarmer** [comment]
Something I did messed up a LaTeX macro.  Since both \ proj and \ amp
occur near the end of the macro list (if you view the html), it is one
of the earlier macros that is the problem.

I looked at the macros, and I suspect it is because \ footnote is defined
in terms of itself.  Try deleting that macro (which is not needed anyway)
and see if that fixes it.

**Comment by Pelonza** [comment]
I fixed the parser errors and got that to work.

I'll look into the macros Wednesday and try to get that processing.

I'm not sure what MBX file you are looking at for "review_matrices" ... the ones I've been pulling from you, which can be seen in the git-hub repository all have "project" in them. Including the most recent pull, which has insights (which wasn't there before). 

[https://github.com/Pelonza/multivariable-calculus-IBL/blob/master/ptx/review_matrices.mbx]

You can see the first instance on line 27

**Comment by davidfarmer** [comment]
I pulled upstream master and now agree that "project" appears 4 times.
Fixing it is no big deal, because it only occurs 4 times, but I'd like
to figure out how it happened.

I assume you saw Rob's posting about waiting to fix the hint problem?

**Comment by Pelonza** [comment]
That worked. Seems to be rendering fine now.. as do the figures now that I've correctly called the generating function...

I guess we wait on "tasks" to be implemented by Rob for fixing the hints/style of the whole document.
I think this means I actually CAN work on figuring out how to include sage/webworks directly and will just manually merge the files that I succeed in changing. (should just be copy/paste into the correct spots in the mbx files)

**Comment by davidfarmer** [comment]
Good.  Change anything you want, and I'll pull before fixing hints
(when Rob gives the green light).

Also let me know if other fixes can be scripted.

**Comment by Pelonza** [comment]
Ok. At this point I'm planning to modify directly the various mbx files... hopefully that won't cause an issue with going through and replacing/fixing the tasks when Rob gives the green light.

As I said, most of my changes should be just additional materials... though who knows, every time i read through the text I find little errors :l

**Comment by Pelonza** [comment]
There does seem to be a potential issue that can be script-fixed.
It seems in generating the mbx files, instead of inserting "tabs" (or 4 spaces) for all the indentations ...it inserted 2 spaces...

This makes editing/inserting new markdown sort of wonky.

**Comment by davidfarmer** [comment]
If you want all the indents to be multiples of 4 spaces, then
I can do that.  But I am not sure why you would think it is better
to have everything shifted over by that much more.

Note that equating tab to 4 spaces is not universal, and that should
be configurable in your editor.

Also, whenever I script a change, I run a postprocesing script that
adjusts the indentation.  I can make it use 2, or 4, or any other
multiple.

I can send you a pull request with 4 spaces as the multiple,
if you want to compare.  But please don't choose to use 4 just
because your editing program seems to expect it.

**Comment by Pelonza** [comment]
Fair enough. I had actually thought a little on that. Go ahead and leave it
at 2, and I'll mess around with sublime to change the default tab spacing...

I think for the markdown with as many indents as may well happen, a bit
less might be far better.

My python code is far more readable with 4 though (which is also the
style-standard for python code).

**Comment by davidfarmer** [comment]
I do a lot of python programming, and try to follow the
standards, including 4 spaces per level.  But for PTX, especially
if you have WeBWorK or other embedded code, 4 spaces moves you
absurdly far to the right.  (In particular, it is not feasible
to use 4 spaces in PTX and also limit to 75 characters per line.)
That is why I decided to use 2 spaces.
But my prettyprinter has that indentation as a parameter.

I am planning to write a PTX analogue of the PEP8 python style
guide.

**Comment by Pelonza** [comment]
Ok. I'm going to close this running issue... 

It looks like everything in it is "done"... except maybe implementing smaller indents etc. Which I'm not sure I have sublime properly set-up for anyway. I'm going to start combing over the pages carefully now in preparation for the semester....so hopefully I can catch/fix anything small that is still broken.
